rmc-docs/src/tutorial-kinds-of-failure.md

```diff
@@ -0,0 +1,161 @@
+
+1. Exercise: Try to run RMC on the above, to see what this kind of failure looks like.
+
+Rust also performs runtime safety checks for integer overflows, much like it does for bounds checks.
+Consider this code (from `tests/overflow.rs`):
+
+```rust
+{{#include tutorial/kinds-of-failure/tests/overflow.rs:code}}
+```
+
+A trivial function, but if we write a property test for it, we immediately find inputs where it fails, thanks to Rust's dynamic checks.
+Here's the output from RMC:
+
+```
+# rmc tests/overflow.rs
+[...]
+** Results:
+./tests/overflow.rs function simple_addition
+[simple_addition.assertion.1] line 6 attempt to compute `move _3 + move _4`, which would overflow: FAILURE
+[simple_addition.overflow.1] line 6 arithmetic overflow on unsigned + in var_3 + var_4: FAILURE
+
+** 2 of 2 failed (2 iterations)
+VERIFICATION FAILED
+```
+
+Notice the two failures: the Rust-inserted overflow check (`simple_addition.assertion.1`) and RMC's explicit overflow check (`simple_addition.overflow.1`).
+
+> **NOTE:** You could attempt to fix this issue by using Rust's alternative mathematical functions with explicit overflow behavior.
+For instance, instead of `a + b` write `a.wrapping_add(b)`.
+>
+> However, [at the present time](https://github.com/model-checking/rmc/issues/480), while this disables the dynamic assertion that Rust inserts, it does not disable the additional RMC overflow check.
+> As a result, this currently still fails in RMC.
+
+</quote>
+
+### Exercise: Classic overflow failure
+
+One of the classic subtle bugs that persisted in many implementations for a very long time is finding the midpoint in quick sort.
+This often naively looks like this (from `tests/overflow-quicksort.rs`):
+
+```rust
+{{#include tutorial/kinds-of-failure/tests/overflow-quicksort.rs:code}}
+```
+
+RMC immediately spots the bug in the above code.
+
+1. Exercise: Fix this function so it no longer overflows.
+(Hint: depending on which approach you take, you may need to add the assumption that `high > low` to your proof harness.
+Don't add that right away, see what happens if you don't. Just keep it in mind.)
+2. Exercise: Prove your new implementation actually finds the midpoint correctly by adding an assertion to the test harness.
+
+## Future work
+
+RMC notably does not check the following:
+
+1. Concurrency bugs, deadlocks, or data races.
+It's possible RMC may be extended in the future to find such issues.
+
+2. Rust type invariants.
+For example, it's undefined behavior in Rust to produce a value of type `bool` that isn't `0` or `1`.
+RMC will not spot this error (in presumably unsafe code), yet.
+
+3. Fully generic functions.
+To write a proof harness and call functions, they must be fully "monomorphized."
+This means we can't currently check a generic function (`foo<T>`) generically.
+Proof harnesses have to be written specializing type parameters (`T`) to concrete types (e.g. `u32`), and check those instead.
```

Comment:
I'd prefer to link to the dashboard so this information doesn't become stale

Reply:
I tried to stick to bigger ideas here than what the dashboard really shows. Set people's expectations appropriately.

Reply:
Ok, in that case, could we make the statement "does not currently check"?